internal/server/middleware.go

```diff
@@ -49,63 +50,80 @@ func DatabaseMiddleware(db *gorm.DB) gin.HandlerFunc {
 	}
 }
 
-func DestinationMiddleware() gin.HandlerFunc {
-	return func(c *gin.Context) {
-		uniqueID := c.GetHeader("Infra-Destination")
-		if uniqueID != "" {
-			destinations, err := access.ListDestinations(c, uniqueID, "", &models.Pagination{})
-			if err != nil {
-				return
-			}
+func updateDestinationLastSeenAt(c *gin.Context) error {
+	uniqueID := c.Request.Header.Get("Infra-Destination")
+	if uniqueID == "" {
+		return nil
+	}
 
-			switch len(destinations) {
-			case 0:
-				// destination does not exist yet, noop
-			case 1:
-				destination := destinations[0]
-				// only save if there's significant difference between LastSeenAt and Now
-				if time.Since(destination.LastSeenAt) > time.Second {
-					destination.LastSeenAt = time.Now()
-					if err := access.SaveDestination(c, &destination); err != nil {
-						sendAPIError(c, err)
-						return
-					}
-				}
-			default:
-				sendAPIError(c, fmt.Errorf("multiple destinations found for unique ID %q", uniqueID))
-				return
+	// TODO: use GetDestination(ByUniqueID())
+	destinations, err := access.ListDestinations(c, uniqueID, "", &models.Pagination{})
+	if err != nil {
+		return err
+	}
+
+	switch len(destinations) {
+	case 0:
+		// destination does not exist yet, noop
+		return nil
+	case 1:
+		destination := destinations[0]
+		// only save if there's significant difference between LastSeenAt and Now
+		if time.Since(destination.LastSeenAt) > time.Second {
+			destination.LastSeenAt = time.Now()
+			if err := access.SaveDestination(c, &destination); err != nil {
+				return fmt.Errorf("failed to update destination lastSeenAt: %w", err)
 			}
 		}
+		return nil
+	default:
+		return fmt.Errorf("multiple destinations found for unique ID %q", uniqueID)
+	}
+}
 
-		c.Next()
+// TODO: remove duplicate in access package
```

Comment:
Planning on addressing this in follow up PRs, but I wanted to avoid exporting it for now since that will change another 20+ places.